[Soap] Step2 회원가입  , Step 3 다양한 컨텐츠 타입

Miensoap opened the pull request:
## 구현 내용

- request handler run 메서드 하위 메서드 분리 : run에서는 분기 처리 , 로그 
- 정규식으로 어떤 요청인지 확인
- create 요청이 들어오면 파라미터 정보로 User 만들어 저장하고 index.hmtl 을 응답하도록 구현
-  >정규식으로 각 파라미터 파싱해 사용

- 파일 이름 없이 경로만 요청이 들어오면 그 경로의 index.html 을 응답
 
## 고민 사항

- thread pool 개수 선택 : 단순히 논리 프로세서 수가 16이라 선택 , 몇 개가 적당한지 고민
- 브라우저에서 같은 요청을 여러 번 보내 , create 메서드가 여러 번 실행되는 문제

- request , response 객체로 만들어 사용 고민 -> Step 3에서 수정

## 기타

- 에러 응답 메서드 통합 예정 
- rebase가 잘못되어 Step1의 커밋들이 새 커밋이 됨

Miensoap committed:
회원 가입 form 제출 : /create 요청

- 정규식으로 create 요청인지 구분

Miensoap committed:
요청 파라미터로 User 생성해 DataBase에 add 구현

- 동일한 요청이 여러 번 들어와서 똑같은 User가 많이 생기는 문제

Miensoap committed:
회원 가입 후 , 메인 페이지로 이동

Miensoap committed:
회원 가입 후 , 메인 페이지로 이동 : 리디렉트 302 found 코드로 변경

godrm commented:
우선 main 브랜치로 보낸 것 같으니 수정해주세요

Miensoap committed:
Request , Response 클래스 추가 , run 의 하위 메서드 나누어 처리

- 이메일 형식 검사 html 상에서 하도록 수정

Miensoap commented:
> 우선 main 브랜치로 보낸 것 같으니 수정해주세요

수정했습니다!

godrm commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,58 +1,110 @@
+            String[] request = requestHeader.split(" ");
 
-        } catch (IOException |  ArrayIndexOutOfBoundsException e) { // 리퀘스트 읽을 때 발생 예외
+            if (createUserHeader.matcher(requestHeader).matches()) {
+                log.info("create request " + requestHeader);
+                createUser(dos, requestHeader);
+            } else if (getFileHeader.matcher(requestHeader).matches()) {
+                log.info("getFile request : " + requestHeader);
+                responseGetFile(dos, request);
+            } else {
+                log.error(requestHeader);
+            }
+        } catch (IOException | ArrayIndexOutOfBoundsException e) { // 리퀘스트 읽을 때 발생 예외
```
이렇게 생각해보세요. RequestHandler는 장군이고 createUserHeader, getFileHeader, responseGetFile 등은 하위에 부하들입니다. 장군이 모든 부하들에게 개별적으로 명령하지 않자나요. 장군은 "공격하라" 한 마디만 하면 그 밑에 장교가 역할별로 다른 행동을 하고, 그 밑에 부하들이 각자 역할에 따라 다른 공격을 하게 되죠. 
여기 코드와 처리도 그런 맥락에서 RequestHandler 가 여러 일을 처리하지말고, 단계적으로 처리하도록 구현해보세요

Miensoap replied:
상위 메서드인 run() 에서 makeRequest , sendResponse 메서드를 호출하고 , 
하위에 있는 Request , Response 클래스에서 각각 판단해 , private 메서드로 다른 역할을 하도록  수정했습니다

godrm commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,58 +1,110 @@
     }
 
-    private byte[] responseBodyFile(File file) throws IOException {
+    private static void createUser(DataOutputStream dos, String requestHeader) throws IOException {
+        // create User 메서드
+        Pattern param = Pattern.compile("=\\w+(%40\\w+\\.com)?");
+        Matcher matcher = param.matcher(requestHeader);
+        List<String> createParameters = new ArrayList<>();
+
+        while (matcher.find()) {
+            createParameters.add(matcher.group());
+        }
+
+        String id = createParameters.get(0).substring(1);
+        String password = createParameters.get(1).substring(1);
+        String nickname = createParameters.get(2).substring(1);
+        String email = createParameters.get(3).substring(1);
+
+        User user = new User(id, password, nickname, email);
+        Database.addUser(user);
+
+        dos.writeBytes("HTTP/1.1 302 Found \r\n");
+        dos.writeBytes("Location:" + "/index.html");
+        dos.writeBytes("\r\n");
+
+        log.info("created : " + Database.findUserById(id).toString());
+    }
```
static 메소드로 구현했다는 것은 RequestHandler 타입과 연관성이 떨어진다는 의미일 수 있습니다
그렇다면 다른 클래스에 메소드로 구현해도 된다는 의미구요. 위에서 언급한 것처럼 이 메소드가 포함되어야 할 클래스를 고려해서 분리해보세요

Miensoap replied:
Request , Response 클래스를 추가해 각각 역할을 나누어 주었습니다.

godrm commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,58 +1,110 @@
+    }
+
+    private static byte[] responseBodyFile(String url) throws IOException {
+        File file = new File("src/main/resources/static" + url);
```
경로는 절대 경로와 상대 경로가 있습니다. 이렇게 하드코딩하게 되면 다른 컴퓨터나 경로 위치에 따라 동작하지 않을 수 있다는 얘기가 됩니다. 이런 경우는 어떻게 처리해야 할까요?

Miensoap replied:
프로젝트 루트 기준 상대 경로로 수정해 보았습니다

honux77 replied:
빌드 도구를 사용하면 그것보다 더 좋은 방법도 있을 수 있는데 한 번 더 찾아서 공부해 보세요.

godrm commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -72,12 +124,12 @@ private void response200Header(DataOutputStream dos, int lengthOfBodyContent, St
+        dos.writeBytes("Content-Length: " + errorMessage.length() + "\r\n");
+        dos.writeBytes("\r\n");
+        dos.writeBytes(errorMessage); // body
     }
```
이전에 피드백 한 것 같은데 responseBody, response200Header, response404 같은 static 메소드들이 하드코딩되어 있고 전혀 HTTP Request와 Response를 대응하는 구조로 되어있지 않습니다. 이 부분을 별도로 분리해보세요

Miensoap replied:
Response 클래스로 역할을 분리한 뒤 , handler에서는 sendResponse() 로 사용하도록 했습니다.
`static 메소드들이 하드코딩되어 있고 전혀 HTTP Request와 Response를 대응하는 구조로 되어있지 않습니다`
피드백이 이 부분을 말씀하신 게 맞을까요?

+ 상태 관련 메서드들은 필요한 정보가 각각 다른 것 같아 고민중입니다.

godrm replied:
네 구조적으로 동일한 데이터와 변경되는 데이터를 하드코딩으로 남겨놓기보다는 데이터 구조 또는 클래스로 표현하려고 노력해보세요

godrm reviewed (changes requested):
앞서 개선사항과 함께 읽어보시고 역할과 책임을 좀 더 나눠보세요
빠르게 구현해야 하는 것이 목표가 아니라, 더 구조적으로 표현해보고 계속해서 개선할 수 있도록 구현하는 방법을 고민해봐야 합니다.
머지하지 않고 열어둘테니 여기 브랜치에 수정하고 커밋하고 푸시하세요

Miensoap committed:
Request 테스트 코드 작성 , 파일 경로 상대경로로 수정

Miensoap committed:
Merge branch 'Step3' into step2

Miensoap committed:
상대 경로 수정 : 프로젝트 기준 상대경로를 사용해야함

Miensoap commented:
**<리뷰 반영>**
- handler와 Response , Request 분리를 Step3 브랜치에서 작업해 , 머지하고 푸시했습니다.
- 상대 경로 사용하도록 수정했습니다.
- 마지막 상태 관련 메서드 통합

Miensoap committed:
헤더 메서드 3개 하나로 통합

- 공통된 부분을 합치기는 했는데 , if문이 추가되고 억지로 합쳐진 느낌이 듬

Miensoap committed:
Response 클래스에서 직접 로그 출력하도록 변경 , main.min.css NoSuchFile 로그 처리하도록 수정

Miensoap committed:
main.min.css 수정하며 생긴 registration이 처리되지 않던 오류 수정

Miensoap committed:
학습 계획 작성

Miensoap committed:
예외 처리 , 로그 수정

- 예외를 미리 catch해서 404 페이지를 응답하지 않던 오류 수정
- 제거했던 New Client Connect 로그 다시 추가
- 요청 로그 형식 보기 좋게 수정

Miensoap committed:
response 테스트 작성

- createUser 테스트 작성
- getFile 은 어떻게 테스트?

godrm commented:
머지가 안되고 있다고 다음 단계꺼를 한꺼번에 올리기 보다는 다른 브랜치에서 작업하는 게 좋습니다
이렇게 되면 특정한 범위를 리뷰하다가, 새로운 기능 범위까지 리뷰를 해야 하기 때문에 현업에서는 아예 머지가 안되는 상황이 됩니다
PR 단위는 의미있는 기능 단위로 리뷰하고 머지하는 것이 중요하거든요.

Miensoap committed:
Request 멤버 중 필수인 method , url 은 생성자로 옮김 , 로그 변경에 따라 테스트 코드 수정

godrm commented on src/main/java/model/User.java:
```diff
@@ -13,6 +15,15 @@ public User(String userId, String password, String name, String email) {
         this.email = email;
     }
 
+    public static User makeUser(List<String> params){
+        String id = params.get(0);
+        String password = params.get(1);
+        String nickname = params.get(2);
+        String email = params.get(3);
+
+        return new User(id, password , nickname , email);
```
makeUser는 전형적인 팩토리 메소드네요. 
이럴 경우에도 배열로 넘기는 것은 좋은 구현이 아닙니다. makeUser()를 호출하는 다른 개발자가 실수할 가능성이 매우 높죠. params가 [] 비어있거나, 1개 2개 100개일 때마다 다 다르게 동작할 수 있습니다. 예외처리가 코드가 많아질 수 밖에 없는 방식이 됩니다.

Miensoap replied:
모두 필수인 속성이라면 파라미터 4개로 만드는 것이 낫겠네요!

godrm replied:
모두 필수 속성이라면 기본 생성자랑 동일해질 겁니다 ㅎㅎ

godrm commented on src/main/java/webserver/Request.java:
```diff
@@ -0,0 +1,98 @@
+        this.url = url;
+    }
+
+    public static Request makeRequest(InputStream in) throws IOException {
+        String[] requestHeader = parseRequest(in);
+        Request request = new Request(requestHeader[0] , requestHeader[1]);
+
+        if (createUserHeader.matcher(request.url).matches()) {
+            request.reqDetail = "createUser";
+            setParams(request);
+        }
+        else if (getFileHeader.matcher(request.url).matches()) {
+            request.reqDetail = "getFile";
+            setFileReqUrl(request);
+        }
+        return request;
+    }
```
Request 같은 클래스는 InputStream 같은 java.io 패키지에 의존하지 않는 게 좋습니다
이게 꼭 InputStream에 의해서만 생성할 지를 여기서 의존하도록 만들면 Request만 생성해서 활용하는 곳도 꼭 InputStream만 사용해서 생성해야 하니까요
이런 메소드는 InputStream을 써야 하는 곳에서 따로 구현하는 게 좋습니다.

Miensoap replied:
실제로 테스트 코드를 작성할 때 InputStream을 사용해 생성해야 하는 일이 있었습니다.

handler에서 헤더 , 바디 등을 미리 가공해 makeRequest 에 주는 것이 나을까 싶습니다

godrm replied:
Response와 마찬가지인데 Request가 꼭 InputStream에서만 생성되야 하는 건 아니죠
그래서 Request 클래스에서 InputStream에 의존할 필요가 없습니다. 
Request를 InputStream으로 생성하는 게 필요한 곳에서 구현하면 되는거죠

godrm commented on src/main/java/webserver/Request.java:
```diff
@@ -0,0 +1,98 @@
+            createParameters.add(matcher.group().substring(1)); // '=' 제거
+        }
+        request.params = createParameters;
+    }
```
이 메소드는 좀 이상한 것 같아요. Request 클래스 메소드인데 static 이라서 멤버 속성을 전혀 다루지 않네요. 인스턴스 메소드여야 하지 않을까요?

godrm commented on src/main/java/webserver/Response.java:
```diff
@@ -0,0 +1,70 @@
+    private static final Logger log = LoggerFactory.getLogger(Response.class);
+    private Response(){}
+    public static void sendResponse(DataOutputStream dos, Request request) throws IOException {
+        if (request.getReqDetail().equals("createUser")) {
+            createUser(dos, request.getParams());
+        } else if (request.getReqDetail().equals("getFile")) {
+            responseGetFile(dos, request.getUrl(), request.getFileType());
+        }
```
Request의 private 속성을 getReqDetail으로 가져와서 비교하는 코드를 이렇게 다른 곳에서 처리하는 게 아니라
Request에게 비교하는 역할을 메소드를 구현하고 값을 전달하는 구조로 바꿔보세요
그래야 getter를 사용하지 않고 객체가 동작을 처리하는 메소드가 많아집니다

Miensoap replied:
Response 에 Request를 넘겨준 후 , 요청에 따른 동작을 메서드와 매핑해서 startsWith(key) 로 구분 후 실행하도록 했습니다

Miensoap commented:
> 머지가 안되고 있다고 다음 단계꺼를 한꺼번에 올리기 보다는 다른 브랜치에서 작업하는 게 좋습니다 이렇게 되면 특정한 범위를 리뷰하다가, 새로운 기능 범위까지 리뷰를 해야 하기 때문에 현업에서는 아예 머지가 안되는 상황이 됩니다 PR 단위는 의미있는 기능 단위로 리뷰하고 머지하는 것이 중요하거든요.

Change request 된 부분만 머지해 올리고 , 나머지는 다음 단계 브랜치에서 작업했어야 했나보네요!
머지한 후에 이제 하나의 브랜치라고 생각해 다음 단계 브랜치를 지워버리고 여기로 통합해 버려 문제가 생긴 것 같아요
적당한 PR 단위에 대해서도 생각해 보겠습니다

godrm commented on src/main/java/webserver/Response.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private static byte[] getFileBytes(String url) throws IOException {
+        File file = new File("./src/main/resources/static" + url); // 프로젝트 기준 상대 경로
```
여기 상대 경로도 시작할 때 전달할 수 있도록 지정하면 어떨까요?
static 파일이 꼭 저기에 있어야만 실행된다는 건 클래스 파일 위치와 함께 고정되는 것이라서요

godrm commented on src/test/java/webserver/RequestTest.java:
```diff
@@ -0,0 +1,45 @@
+        assertThat(request.getParams().get(1)).isEqualTo("test");
+    }
+
+    private static Request makeRequest(String url) throws IOException {
+        InputStream inputStream = new ByteArrayInputStream(url.getBytes());
+        return Request.makeRequest(inputStream);
+    }
```
이렇게 makeRequest를 반복해야 한다면 fixture를 사용할 수 있지 않을까요?

godrm commented on src/test/java/webserver/ResponseTest.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Test
+    @DisplayName("createUser 요청에 대해 매개변수가 알맞게 파싱되고 , Request 객체가 반환되어야 한다")
+    void createUserResponse() throws IOException {
+        String url = "GET /create?userId=test&password=test&name=test&email=test%40naver.com";
+        Request request = makeRequest(url);
+
+        Response.sendResponse(dos,request);
+        assertThat(Database.findUserById("test")).isNotNull();
```
Response 테스트할 때 DataOutputStream 의존성을 제외하고 Response 클래스만 테스트할 수 있도록 개선해보세요

Miensoap replied:
Request 와 같은 맥락으로 , Response 테스트를 작성할 때도 OutputStream이 필요한 일이 생겼습니다.

OutputSream에 쓰기 할 데이터를 만드는 것 까지만 Response가 하도록 하면 어떨까 싶습니다

godrm replied:
아뇨 분리하는 게 더 좋습니다. Response를 OutputStream에 쓰도록 결정하는 것은 그 상위에서 할 일이거든요. 서로 연관이 없는 구조가 더 좋습니다.

Miensoap replied:
Response 내에서 단계적으로 데이터 만들기 -> 쓰기 하도록 수정해서 
만들어진 데이터를 테스트 할 수 있도록 해보겠습니다

Miensoap committed:
Request 생성 시 inputStream 이 아니라 , 문자열 만으로 생성 가능하게 변경 , 그에 따른 테스트 코드 변경

Miensoap committed:
User 생성 파라미터 변경 -> 생성자와 동일해 생성자 사용

Miensoap committed:
Response 생성과 dos 분리 , 단계적으로 response header , body 를 만들고 send() 에서 dos에 쓰기

Miensoap committed:
Response 테스트 오류 수정 : body 비교인데 header 사용 중인 오류 수정

godrm commented on src/main/java/webserver/Request.java:
```diff
@@ -1,98 +1,48 @@
-    private List<String> params;
+    private final String method;
+    private String path;
+    private Map<String, String> params;
```
오호 java.io 패키지 의존도를 제거하셨네요
params 라는 것도 HTTP 용어로 표현하면 좋을 것 같습니다. 전혀 다른 의미가 될 수 있거든요

Miensoap replied:
어제 공부하며 알게되었는데 , url 의 구성 요소 중  query - query parameter - query string 으로 부르더라구요 , 
query 가 더 알맞은 표현일 것 같네요

godrm commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -17,13 +18,37 @@ public void run() {
         }
     }
+
+    private static Request makeRequest(String requestMessage) {
+        String[] requestHeaderFirst = requestMessage.split("\r\n")[0].split(" ");
+        Request request = new Request(requestHeaderFirst[0], requestHeaderFirst[1]);
+        log.info("Request : " + request.getLog());
+        return request;
+    }
```
static 메소드가 있을 때는 우선 인스턴스 메소드로 표현가능한가? 어떤 클래스에 구현하면 될까? 어떤 역할인가? 고민해보는 게 좋습니다
RequestHandler는 Request를 생성하는 역할보다는 하위 다른 객체들을 연결하고 조합하는 역할에 집중하면 좋을 것 같습니다. makeRequest()는 엄밀하게 구분해보면 Request의 생성자 또는 팩토리 메소드가 될 것 같네요.

Miensoap replied:
후에 request 메시지의 전체 header와 ,  body 부분을 읽도록 변경하면, header와 body 정보를 조합해 
온전한 Request 객체를 만들어주는 메소드가 될 것 같아요. 

이 경우에도 두 정보를 모두 Request 클래스 내에서 처리한다면 생성자 또는 팩토리 메소드 일 것 같은데 , 

Header , Body 등으로 하위 객체들을 분리하는게 나을것 같다는 의미로 이해했습니다.
제가 맞게 이해했을까요?

godrm commented on src/main/java/webserver/Response.java:
```diff
@@ -9,62 +9,98 @@
+    }
 
-        log.info(request.getLog() + " Complete");
+    public void sendResponse(DataOutputStream dos) throws IOException {
+//        log.info(new String(header));
+//        log.info(new String(body));
+        dos.write(header, 0 , header.length);
+        dos.write(body, 0 , body.length);
+        dos.flush();
     }
```
sendReponse()도 이렇게 생각해보세요
다른 객체에서 Reponse 객체를 만들고 어디로 보낼지는 response 인스턴스가 알지 못합니다. 네트워크로 전송할지, 파일에 저장할지, OutputStream에 write할지는 여기서 처리하지 않는 게 좋습니다. 여기 의존성이 남아있어서 아쉽네요

Miensoap replied:
이전 코멘트에서도 질문했던 내용인데 , 답변을 잘못 이해한 것 같네요
response 메시지를 작성하는 주체와 send 하는 주체를 분리해 보겠습니다

godrm commented on src/main/java/webserver/WebServer.java:
```diff
@@ -11,6 +11,7 @@
 public class WebServer {
     private static final Logger logger = LoggerFactory.getLogger(WebServer.class);
     private static final int DEFAULT_PORT = 8080;
+    public static final String staticSourcePath = "./src/main/resources/static";
```
이런 포트나 경로는 호눅스가 언급한 것처럼 빌드나 배포 과정에서 결정되는 경우가 많습니다
하드코딩을 하더라도 상수보다는 외부에서 지정해주는 변수 형태를 고려해보셔도 좋습니다

Miensoap replied:
config.yaml 등의 파일에서 설정하는 방법은 어떤가요?

자바 파일 내의 String을 수정하는 것과 config 파일을 수정하는 것의 차이와 이점에 대해 아직 잘 감이 오지 않습니다
똑같이 문자를 수정해야 하는 것 아닌가 생각이 들어서요

godrm commented on src/test/java/webserver/RequestTest.java:
```diff
@@ -5,41 +5,34 @@
         Request request = makeRequest(url);
-        assertThat(request.getLog()).isEqualTo("CREATEUSER  :  /create?userId=test&password=test&name=test&email=test%40naver.com");
-        assertThat(request.getParams().get(1)).isEqualTo("test");
+        assertThat(request.getParams().get("userId")).isEqualTo("test");
+        assertThat(request.getParams().get("password")).isEqualTo("test");
+        assertThat(request.getParams().get("name")).isEqualTo("test");
+        assertThat(request.getParams().get("email")).isEqualTo("test%40naver.com");
```
테스트 코드를 작성하면서 좀 어색하지 않았나요? 
값 하나를 비교하는 데 전체를 다 가져오는 게 반복됩니다. getParams() 대신에 아예 header 값을 넘겨주고 하나만 가져오는 건 어떨까. 테스트 코드를 보면서도 알수 있습니다
테스트 코드를 단지 동작을 확인하는 게 아니라, 해당 API를 사용하는 관점에서 불편함은 없는지, 합리적인지, 효율적인지도 같이 살펴보실 수 있어야 합니다.

Miensoap replied:
getParams 중복이 어색하다는 의미로 이해했습니다. 
그 다음 get에 대한 질문인데 , 
혹시 params 중 일부만 테스트하는 건 어떤가요? 그렇게 되면 테스트의 의도에 맞는 검증이 부족할까요?

godrm reviewed (approved):
거의 대부분 개선이 된 것 같은데 여전히 일부 아쉬운 부분이 있어서 언급했습니다.
마지막 단계를 아마 대부분 구현을 해놓은 상태일테니 마무리하면서 리팩터링도 해보세요 
수고하셨습니다
